### PR TITLE
Add initial Qt backend implementation

### PR DESCRIPTION
**Pull request by Ksengine**
Basics only.
Qt import included in `lib` folder.
this is a startup for all developers.
**work in progress**

## PR Checklist:
- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by goanpeca** [comment]
Hi @Ksengine thanks for working on this.

How about using some of the supported shims that are out there and are already taking care of multiple versions of Qt and PyQt/pyside?

Like  https://pypi.org/project/QtPy/

Disclaimer: I am/was one of the core devs of that one.

**Comment by Ksengine** [comment]
I had recently created some patches.

**Comment by Ksengine** [comment]
I know about QtPy.
we will ask maintainers to add QtPy as dependency.
**is QtPy supports for PySide2, PyQt5, PySide,  PyQt4**
Thank you:exclamation:

**Comment by freakboy3742** [comment]
I might need someone to give an explanation for the design decision here. I know there are multiple bindings for Qt; why would we not just pick *one* and use it? 

* PyQT is GPL/commercial, which essentially means it's toxic from the perspective of our project. 
* PySide is the official binding for Qt4
* Pyside2 is the official binding for Qt5

To my eyes, that means we should be targeting Pyside2. It's LGPL, and it's an official binding of the most recent development branch. What am I missing? What is the virtue of using an abstraction layer like QtPy in this context?

I see the value in using an abstraction layer as an external API - but in our case, Toga *is* the abstraction API. Writing an abstraction API that uses an abstraction API seems like an extra layer that isn't really needed. Am I missing something about the Qt landscape here?

**Comment by goanpeca** [comment]
I think supporting PySide2 should be enough.

>I see the value in using an abstraction layer as an external API - but in our case, Toga is the abstraction API. Writing an abstraction API that uses an abstraction API seems like an extra layer that isn't really needed. Am I missing something about the Qt landscape here?

Well the APIs on PySide/PyQt4, PySide2/Pyqt5 change slightly for some methods, and some imports, the shim makes them uniform and uses PyQt5 (mostly) as the layer to access everything else.

However probably supportuing only pyside2 makes more sense for toga, since it is the official non GPL binding.

**Comment by Ksengine** [comment]
> I had recently created some patches.

I can make the abstraction layer if you won't to use another abstraction layer.
We'll add support for PyQt5, PySide2, PyQt4 and PySide2.
PyQt is installed on some users machines.
Users can choose one, warn them about licences (in docs)
API is like PyQt5(approximately equal to  PySide2)
Only add patches for needed API methods.

**Comment by Ksengine** [comment]
**toga is a basic gui framework yet. It needs more development**

**Comment by freakboy3742** [comment]
@Ksengine We can afford to be opinionated about API usage. We don't need to add (and maintain) support for 4 separate Qt APIs, or support an abstraction layer that itself supports 4 APIs. The "PyQt is already installed on some machines" argument doesn't hold here. BeeWare is designed to build standalone apps, so what is currently installed on a user's machine is of no relevance. "pip install toga" should Just Work, and shouldn't require any choices on behalf of the user, or dependence on installed software.

So - unless you can suggest a compelling reason to choose otherwise, I'd suggest a PySide2 dependency is all we should aim for.

**Comment by Ksengine** [comment]
what about PyQt5

**Comment by freakboy3742** [comment]
What about it? 

As I said earlier, PyQt5 is GPL (which makes it toxic from the perspective of the project), and it's an unofficial binding. Is there a reason to *not* use PySide2, which is (a) official, and (b) not problematic from a license perspective?

**Comment by Ksengine** [comment]
PyQt5 is chosen when PySide2 not installed.

can you read licenses and reply me about the legal state because my law knowladge is low.
if you want advice you can contact PyQt Creaters(http://riverbankcomputing.com/)

**Comment by freakboy3742** [comment]
PyQt5 won't ever *not* be installed - it will be specified as an install requirement of the Toga-QT backend.

I *have* read the licenses (or rather - I don't need to - because they're standard licenses). PyQT is GPL/commercial. PySide is LGPL. There's a big difference between the two; LGPL is the only one that's viable for a project like Toga, because we can't transitively enforce our users use a commercial license, and we aren't going to change our license from BSD.

So - once again: unless there's a compelling *technical* reason to use PyQt5, I'm saying use PySide2, and don't provide *any* allowance in the backend for an alternative.

**Comment by freakboy3742** [comment]
Sorry - that should be *Pyside2* won't ever *not* be installed. Got myself confused over the naming of the official bindings.

**Comment by Ksengine** [comment]
@freakboy3742 
Ok,
but PyQt5 API is more richer than PySide2

**Comment by freakboy3742** [comment]
@Ksengine Ok, so *that* is a technical reason. 

However, if the original plan was to support PyQt5 *and* Pyside2 (and, based on the current version of this patch, pyside 1 and pyQt4) - how were you planning to overcome the lack of "richness"?

I can't understate how much of a problem GPL is. It's a *major* impediment to adoption, and it has *serious* consequences for Toga and BeeWare as a project. If there is literally no way build a viable backend using PySide (or another non-GPL library), I'd actually be inclined to say "well, we can't officially support Qt". That's how big a deal GPL is.

**Comment by Ksengine** [comment]
> @Ksengine Ok, so that is a technical reason.

no Its a simple reason.
I can understand the effects of GPL.
I will make patches for PySide2
just argued with you to add PyQt5 to Toga because that reason.
Dont mind it.
I will remove PyQt5 imports

**Comment by Ksengine** [comment]
I need help to develop Qt port(for some parts like run asyncio GUI loop)
**Can you help by contributing Qt port**

**Comment by Ksengine** [comment]
**I hope to make PyQt5 port of Toga in my own repo**

File: nursery/qt/toga_qt/libs/utils.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+try:  # try import PyQt5
+    from PySide2.QtCore import qVersion
+    GUI = 'pyside2'
+
+except ImportError:
+    try:  # try import PySide2
+        from PySide.QtCore import qVersion
```
Comment (freakboy3742) [inline nursery/qt/toga_qt/libs/utils.py:7]: Again - this isn't needed. We can *enforce* the availability of PySide2. There no need for a fallback here.

Comment (Ksengine) [reply nursery/qt/toga_qt/libs/utils.py:7]: dont mind just added for pyside1 users

Comment (Ksengine) [reply nursery/qt/toga_qt/libs/utils.py:7]: @freakboy3742  is it ok

Comment (freakboy3742) [reply nursery/qt/toga_qt/libs/utils.py:7]: I don't know how many times I can say it. We can (and will) *enforce* the presence of PySide2. There will be *exactly zero* users of Toga-qt who do not have PySide2 installed. There is *absolutely no use case* for a fallback when you can *guarantee* the presence of your preferred API.

Comment (Ksengine) [reply nursery/qt/toga_qt/libs/utils.py:7]: ok

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Since I've been flagged for a review...

1. There are still references to PySide 1.
2. It's not clear to my why you've defined QtGui, QtWidgets etc whose sole purpose is to import their analogs in PySide2. On GTK, it's needed because you can't just "import gtk". On Cocoa we need to wrap the Objective C APIs. However, if PySide provides a literal API we can use, then there's no reason not to use it directly.

**Comment by Ksengine** [comment]
pyside1 removed

**Comment by Ksengine** [comment]
2. I need to add some patches to pyside2

**Comment by Ksengine** [comment]
I can manage this patches.
for frontend I need help.can you assign some developers(+you if can)

**Review by Ksengine — COMMENTED** [review]
all fixed

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
"Assign some developers"? You volunteered to work on this. You're the developer who is working on it. We don't have a huge pool of developers just waiting to be assigned to projects.

Also - there's not much point asking for another review when you haven't addressed the points from the *first* review yet.

**Comment by Ksengine** [comment]
I had addressed the points from the first review.

**Comment by freakboy3742** [comment]
I don't see how you've addressed point 2. There are still mirror modules in the libs folder, and no sign of the "patches" that you've referenced (or even a satisfactory description of what those patches are)

**Comment by Ksengine** [comment]
added version patches (in `util.py`)
I will add some patches later if they wanted.

**Comment by freakboy3742** [comment]
*If* they are needed, *that* would be the time to add modules. In the meantime, you're imposing an overhead for no demonstrated practical purpose.

**Comment by Ksengine** [comment]
no need to add patches for QtGui, and QtWidgets.
I will delete them from libs folder.

**Comment by Ksengine** [comment]
> _If_ they are needed, _that_ would be the time to add modules. In the meantime, you're imposing an overhead for no demonstrated practical purpose.

I cant understand this line because my english knowledge is low.

**Comment by freakboy3742** [comment]
I'm saying that you've added some modules, but they're not actually doing anything. However, having those modules will make imports slower, because you need to parse, compile, and process them.

That is - that portion of the code has a cost, but no benefit. So why pay the cost? Until there is a reason to have the extra modules there's no point paying the cost.

**Comment by Ksengine** [comment]
ok I will remove them

**Comment by Ksengine** [comment]
in Qt `App` real `QApplication` is assigned to `native` attribute.
can I inherit it.
like 
```python
class App(QApplication):
    def create(self):
        self.native = self

**Comment by freakboy3742** [comment]
The three layers in Toga (interface, impl and native) shouldn't be combined. That means the App class shouldn't be inheriting from QApplication - The App object should have an attribute `native` that is the QApplication instance.

**Comment by Ksengine** [comment]
we can combine impl and native.
then in future we can easily add new api methods.

**Comment by freakboy3742** [comment]
No, we can't. Toga used to have a 2 layer architecture. Impl and native are now separated because of the complications we encountered.

As an aside - Github is not the best forum to have these sort of generic back-and-forth design conversations. I suggest you join the Gitter chat room: https://gitter.im/beeware/general

**Comment by Ksengine** [comment]
for asyncio loop I will use https://github.com/gmarull/asyncqt

**Comment by Ksengine** [comment]
I cant develop this for few months.(**importatnt exam**)

**Comment by freakboy3742** [comment]
It's a little frustrating that you've embarked on a massive project, asked for assistance, and then abandoned the project in the space of less than 2 days, because of an external commitment that I presume you knew about before you started.

Closing this PR for now. If you are able to resume work after your exam, you can reopen it.

**Comment by Ksengine** [comment]
no Its suddenly happened(according to announcement of government).

**Comment by Ksengine** [comment]
I like to develop basics before I leave.

**Comment by Ksengine** [comment]
Hey I am here again. exam is in march( due to COVID-19)

**Comment by freakboy3742** [comment]
I *think* you should have access to re-open the PR; if you don't, let me know and I'll push the button for you.

**Comment by freakboy3742** [comment]
... and it looks like I might have pressed the button.

**Comment by Ksengine** [comment]
> I _think_ you should have access to re-open the PR; if you don't, let me know and I'll push the button for you.

no,I hadnt

**Comment by freakboy3742** [comment]
Closing due to inactivity.

**Comment by johnzhou721** [comment]
Heads up for anyone tracking this (in case subscribers here hasn't subscribed to the issue ticket):  Qt support based on PySide6 has now been added by #3769.

Original ticket #1142; backfill management ticket #3914.  Functionality is being completed rapidly; corranwebster is in the process of implementing most if not all non-hardware functionalities, with a lot of things already completed.

**Pull request closed without merging**